src/contracts/middleware/BLSIndexRegistryCoordinator.sol

```diff
@@ -0,0 +1,163 @@
+            = abi.decode(deregistrationData, (BN254.G1Point, uint32[], uint32));
+        // call internal function to deregister the operator
+        _deregisterOperator(msg.sender, pubkey, quorumToOperatorListIndexes, globalOperatorListIndex);
+    }
+
+    /**
+     * @notice Deregisters the operator from the middleware
+     * @param pubkey is the BLS public key of the operator
+     * @param quorumToOperatorListIndexes is the list of the operator's indexes in the quorums they are registered for in the IndexRegistry
+     * @param globalOperatorListIndex is the operator's index in the global operator list in the IndexRegistry
+     */
+    function deregisterOperator(BN254.G1Point memory pubkey, uint32[] memory quorumToOperatorListIndexes, uint32 globalOperatorListIndex) external {
+        _deregisterOperator(msg.sender, pubkey, quorumToOperatorListIndexes, globalOperatorListIndex);
+    }
+
+    function _registerOperator(address operator, bytes calldata quorumNumbers, BN254.G1Point memory pubkey) internal {
+        // TODO: check that the sender is not already registered
+
+        // get the quorum bitmap from the quorum numbers
+        uint256 quorumBitmap = BytesArrayBitmaps.orderedBytesArrayToBitmap_Yul(quorumNumbers);
+        require(quorumBitmap != 0, "BLSIndexRegistryCoordinator: quorumBitmap cannot be 0");
+
+        // register the operator with the BLSPubkeyRegistry and get the operatorId (in this case, the pubkeyHash) back
+        bytes32 operatorId = blsPubkeyRegistry.registerOperator(operator, quorumNumbers, pubkey);
+
+        // register the operator with the IndexRegistry
+        indexRegistry.registerOperator(operatorId, quorumNumbers);
+
+        // register the operator with the StakeRegistry
+        _registerOperator(operator, operatorId, quorumNumbers);
+
+        // set the operatorId to quorum bitmap mapping
+        operatorIdToQuorumBitmap[operatorId] = quorumBitmap;
+
+        // set the operator struct
+        operators[operator] = Operator({
+            operatorId: operatorId,
+            fromTaskNumber: serviceManager.taskNumber(),
+            status: OperatorStatus.REGISTERED
+        });
+    }
+
+    function _deregisterOperator(address operator, BN254.G1Point memory pubkey, uint32[] memory quorumToOperatorListIndexes, uint32 globalOperatorListIndex) internal {
+        require(operators[operator].status == OperatorStatus.REGISTERED, "BLSIndexRegistryCoordinator._deregisterOperator: operator is not registered");
+
+        // get the operatorId of the operator
+        bytes32 operatorId = operators[operator].operatorId;
+        require(operatorId == pubkey.hashG1Point(), "BLSIndexRegistryCoordinator._deregisterOperator: operatorId does not match pubkey hash");
+
+        // get the quorumNumbers of the operator
+        bytes memory quorumNumbers = BytesArrayBitmaps.bitmapToBytesArray(operatorIdToQuorumBitmap[operatorId]);
```

Comment:
is this operation not very expensive?
I was imagining that we get the bytes array as input here as well, use `bytesArrayToBitmap` again and check against storage.
could be worth comparing, but I suppose this works at least.

Reply:
ah! I wasn't even thinking about that, great catch: https://github.com/Layr-Labs/eigenlayer-contracts/pull/76/commits/9674a65ef7e048e97ade7746317a47e620c05aa6